### PR TITLE
RDP (remote desktop) backend

### PR DESCRIPTION
**Pull request by ddevault**
- [x] Spin up FreeRDP server
- [x] Allocate headless outputs for each peer
- [x] Send frame updates to clients via RFX
- [x] Send frame updates to clients via NSC
- [x] Send raw frame updates for clients which support neither
- [x] Keyboard input
- [x] Pointer input
- [x] Clean up

### CODE REVIEW

**Comment by emersion** [comment]
Would be nice to let the compositor allocate the output, so that it can decide to either create a headless one or use an existing one. That would also remove all the OpenGL stuff from the backend.

**Comment by ddevault** [comment]
Hmm, maybe. The way I saw this working out is that the compositor would mirror to the RDP outputs if they wanted to use an existing one. I also like that this lets you transparently add RDP support without any changes to the compositors.

**Comment by emersion** [comment]
Still unsure whether this should be merged or not. A client with virtual-keyboard + virtual-pointer + virtual-output would probably be better.

**Comment by ddevault** [comment]
I'm starting to wonder if it wouldn't be nice to have both.

**Comment by emersion** [comment]
In any case, you could use a headless backend inside the RDP backend to avoid having to re-implement `wlr_output`.

I still think the compositor should be able to customize the output if needed, this could be optional. This can be useful for mirroring or remote controlling.

**Comment by ddevault** [comment]
That seems like a nice to have for v2 of this patch.

**Comment by emersion** [comment]
Note: this backend is better than Weston's because it uses OpenGL (the Weston backend forces the software renderer).

**Comment by ddevault** [comment]
Or it would be if this didn't force llvmpipe for some reason <_< headless backend has the same problem. No luck figuring it out yet.

**Comment by progandy** [comment]
@ddevault Maybe open a render node fd (```/dev/dri/render...```) and create a normal GBM backed mesa platform just without any crtcs or outputs? My very shallow reading of surfaceless mesa seems to indicate that it is built on top of llvmpipe.

**Comment by ddevault** [comment]
Maybe? I'm not sure how that works to be honest. Interested in trying a patch?

**Comment by emersion** [comment]
>Maybe open a render node fd (`/dev/dri/render...`) and create a normal GBM backed mesa platform just without any crtcs or outputs?

That's how the not-yet-merged Weston patch works: https://lists.freedesktop.org/archives/wayland-devel/2017-July/034423.html

**Comment by emersion** [comment]
I'd first try to read the mesa code though.

**Comment by progandy** [comment]
> Maybe? I'm not sure how that works to be honest. Interested in trying a patch?

Currently not really. I don't really know how that works either. I just remembered reading about render nodes for unprivileged gpu access without a DRM master for computation and DRI_PRIME.

I found this implementation in virgl: https://gitlab.collabora.com/virgl-es/virglrenderer/blob/master/src/virgl_egl_context.c

#1114 is probably related. A single DRM backend that supports real and virtual outputs if it is run as DRM master. If it is started on a render node, it will only have virtual outputs and be completely headless.

**Comment by emersion** [comment]
Other idea: use `EGL_PLATFORM_DEVICE_EXT`. I might try this this weekend.

**Comment by ascent12** [comment]
>use EGL_PLATFORM_DEVICE_EXT

Mesa does not support this.

**Comment by emersion** [comment]
On my machine `EGL_PLATFORM_SURFACELESS_MESA` correctly selects i915 and doesn't fallback to software rendering. Maybe this is a Mesa bug?

Can you try with `export EGL_LOG_LEVEL=debug`, just to make sure we're not missing anything?

**Comment by progandy** [comment]
I walked through the code and it look like surfaceless does try to create a render node: https://gitlab.freedesktop.org/mesa/mesa/blob/master/src/egl/drivers/dri2/platform_surfaceless.c#L358

* first try: render node and hardware rendering
* second try: DRM master with software rendering
* last try: swrast without DRM

I did a log as well, looks like r600 is loaded.

    EGL_LOG_LEVEL=debug WLR_RDP_TLS_CERT_PATH="$PWD/cert/cert.pem" WLR_RDP_TLS_KEY_PATH="$PWD/cert/key.pem" WLR_BACKENDS=rdp ./rootston 2>rootston-rdp.log

[rootston-rdp.log](https://github.com/swaywm/wlroots/files/2921365/rootston-rdp.log)

**Comment by ddevault** [comment]
This is ready for review. It still ends up using llvmpipe on my system, but that can be solved separately.

File: backend/rdp/peer.c [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+		(struct wlr_rdp_peer_context *)client->context;
+	struct wlr_rdp_backend *backend = context->backend;
+	rdpSettings *settings = client->settings;
+
+	if (!settings->SurfaceCommandsEnabled) {
+		wlr_log(WLR_ERROR, "RDP peer does not support SurfaceCommands");
+		return FALSE;
+	}
+
+	context->output = wlr_rdp_output_create(backend, context,
```
Comment (emersion) [inline backend/rdp/peer.c]: This is missing a NULL check

File: backend/rdp/peer.c [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+	}
+
+	context->pointer = wlr_rdp_pointer_create(backend);
+	// Use wlroots' software cursors instead of remote cursors
+	POINTER_SYSTEM_UPDATE pointer_system;
+	rdpPointerUpdate *pointer = client->update->pointer;
+	pointer_system.type = SYSPTR_NULL;
+	pointer->PointerSystem(client->context, &pointer_system);
+
+	context->keyboard = wlr_rdp_keyboard_create(backend, settings);
```
Comment (emersion) [inline backend/rdp/peer.c]: Ditto

File: backend/rdp/peer.c [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+			context->output->wlr_output.height);
+	nsc_context_reset(context->nsc_context,
+			context->output->wlr_output.width,
+			context->output->wlr_output.height);
+
+	if (context->flags & RDP_PEER_ACTIVATED) {
+		return TRUE;
+	}
+
+	context->pointer = wlr_rdp_pointer_create(backend);
```
Comment (emersion) [inline backend/rdp/peer.c]: Ditto

File: backend/rdp/peer.c [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+static int xf_suppress_output(rdpContext *context,
+		BYTE allow, const RECTANGLE_16 *area) {
+	struct wlr_rdp_peer_context *peer_context =
+		(struct wlr_rdp_peer_context *)context;
+	if (allow) {
+		peer_context->flags |= RDP_PEER_OUTPUT_ENABLED;
+	} else {
+		peer_context->flags &= (~RDP_PEER_OUTPUT_ENABLED);
```
Comment (emersion) [inline backend/rdp/peer.c]: Style: extra parentheses here

File: backend/rdp/peer.c [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+	struct wlr_input_device *wlr_device = &context->pointer->wlr_input_device;
+	struct wlr_pointer *pointer = wlr_device->pointer;
+	bool frame = false;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+
+	if (flags & PTR_FLAGS_MOVE) {
+		struct wlr_event_pointer_motion_absolute event = { 0 };
+		int width, height;
+		wlr_output_effective_resolution(
```
Comment (emersion) [inline backend/rdp/peer.c]: `wlr_output_effective_resolution` shouldn't be used here. Instead, `wlr_output->{width,height}` should be used and the input device should be mapped to the output.

(Applies to all absolute motion events)

File: backend/rdp/pointer.c [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+#include <wlr/interfaces/wlr_input_device.h>
+#include <wlr/util/log.h>
+#include "backend/rdp.h"
+#include "util/signal.h"
+
+static struct wlr_input_device_impl input_device_impl = { 0 };
+
+struct wlr_rdp_input_device *wlr_rdp_pointer_create(
+		struct wlr_rdp_backend *backend) {
+	struct wlr_rdp_input_device *device =
```
Comment (emersion) [inline backend/rdp/pointer.c]: This is missing a NULL check

File: backend/rdp/pointer.c [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	const char *name = "rdp";
+	struct wlr_input_device *wlr_device = &device->wlr_input_device;
+	wlr_input_device_init(wlr_device, WLR_INPUT_DEVICE_POINTER,
+			&input_device_impl, name, vendor, product);
+
+	if (!(wlr_device->pointer = calloc(1, sizeof(struct wlr_pointer)))) {
+		wlr_log(WLR_ERROR, "Failed to allocate RDP pointer device");
+		return NULL;
+	}
+	wlr_pointer_init(wlr_device->pointer, NULL);
```
Comment (emersion) [inline backend/rdp/pointer.c]: Right before, you can map the pointer to the output. See e.g. how the Wayland backend does it.

```c
wlr_dev->output_name = strdup(output->wlr_output.name);
```

File: include/wlr/backend/rdp.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+#include <wlr/types/wlr_input_device.h>
+#include <wlr/types/wlr_output.h>
+
+/**
+ * Creates an RDP backend. An RDP backend will create one output, keyboard, and
+ * pointer for each client that connects.
+ * TODO: add freerdp options like bind port to this
+ */
+struct wlr_backend *wlr_rdp_backend_create(struct wl_display *display,
+	wlr_renderer_create_func_t create_renderer_func,
```
Comment (emersion) [inline include/wlr/backend/rdp.h]: Indentation is off here

File: include/wlr/backend/rdp.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef WLR_BACKEND_RDP_H
+#define WLR_BACKEND_RDP_H
+#include <wlr/backend.h>
+#include <wlr/types/wlr_input_device.h>
+#include <wlr/types/wlr_output.h>
+
+/**
+ * Creates an RDP backend. An RDP backend will create one output, keyboard, and
+ * pointer for each client that connects.
+ * TODO: add freerdp options like bind port to this
```
Comment (emersion) [inline include/wlr/backend/rdp.h]: Since there is `wlr_rdp_backend_set_port`, is this still relevant?

File: include/wlr/types/wlr_output.h [hunk 10]
```diff
@@ -277,4 +277,14 @@ enum wl_output_transform wlr_output_transform_invert(
 enum wl_output_transform wlr_output_transform_compose(
 	enum wl_output_transform tr_a, enum wl_output_transform tr_b);
 
+/**
+ * Marks the output as in need of a buffer swap.
+ */
+void wlr_output_update_needs_swap(struct wlr_output *output);
```
Comment (emersion) [inline include/wlr/types/wlr_output.h]: This shouldn't be in this header. There's already a definition of this in `include/wlr/interfaces/wlr_output.h`. This function is intended to be used by backends, not compositors.

File: include/wlr/types/wlr_output.h [hunk 11]
```diff
@@ -277,4 +277,14 @@ enum wl_output_transform wlr_output_transform_invert(
...
+
+/**
+ * Damages the whole output.
+ */
+void wlr_output_damage_whole(struct wlr_output *output);
```
Comment (emersion) [inline include/wlr/types/wlr_output.h]: Ditto

File: backend/meson.build [hunk 12]
```diff
@@ -53,6 +53,22 @@ if logind.found()
 	backend_deps += logind
 endif
 
+if freerdp.found()
```
Comment (emersion) [inline backend/meson.build]: `and winpr2.found()`?

**Comment by ddevault** [comment]
Addressed @emersion's comments and rebased into a single (changelog-ready) commit.

File: backend/rdp/peer.c [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+		rdpInput *input, UINT16 flags, UINT16 x, UINT16 y) {
+	struct wlr_rdp_peer_context *context =
+		(struct wlr_rdp_peer_context *)input->context;
+	struct wlr_input_device *wlr_device = &context->pointer->wlr_input_device;
+	struct wlr_pointer *pointer = wlr_device->pointer;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	struct wlr_event_pointer_motion_absolute event = { 0 };
+	int width, height;
+	wlr_output_effective_resolution(
```
Comment (emersion) [inline backend/rdp/peer.c]: `wlr_output.{width,height}` should be used here

File: backend/rdp/output.c [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+		wlr_log(WLR_ERROR, "Raw updates are not supported; use rfx or nsc");
+	}
+	wlr_output_send_present(wlr_output, NULL);
+	return ret;
+}
+
+static void output_destroy(struct wlr_output *wlr_output) {
+	struct wlr_rdp_output *output =
+		rdp_output_from_output(wlr_output);
+	wl_event_source_remove(output->frame_timer);
```
Comment (emersion) [inline backend/rdp/output.c]: This will segfault if `output->frame_timer` is NULL (not yet initialized).

Backtrace: https://gist.github.com/kennylevinsen/54cb7d55c1ec1d7e86d84b11573e0d87

File: backend/rdp/output.c [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+static EGLSurface egl_create_surface(struct wlr_egl *egl, unsigned int width,
+		unsigned int height) {
+	EGLint attribs[] = {
+		EGL_WIDTH, width,
+		EGL_HEIGHT, height,
+		EGL_RED_SIZE, 1,
+		EGL_GREEN_SIZE, 1,
+		EGL_BLUE_SIZE, 1,
```
Comment (emersion) [inline backend/rdp/output.c]: Actually, we already ask for red/green/blue in `backend.c`, and these are invalid for `eglCreatePbufferSurface`. Can you revert this change?

**Comment by emersion** [comment]
Thanks!

**Comment by RorraVox** [comment]
Can someone explain in simple words how start freerpd with sway? I use remmina i think should be compatible.
I cant find nothing about that...
Thank you.

**Comment by twhitehead** [comment]
The original patch put a bunch of info into *docs/env_vars.md*. Looks like it was dropped in ebdbe177d62d3c8bab04f77b86326804ff185bef in favour of [wayvnc](https://github.com/any1/wayvnc) though.
